src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,442 @@
+            {
+                (int id, int index) = DeconstructSlot(containingSlot);
+                var variables = GetVariablesForId(id)!;
+                var variableBySlot = variables._variableBySlot;
+                for (index++; index < variableBySlot.Count; index++)
+                {
+                    var variable = variableBySlot[index];
+                    if (variable.ContainingSlot == containingSlot)
+                    {
+                        builder.Add((variable, ConstructSlot(id, index)));
+                    }
+                }
+            }
+
+            private Variables GetVariablesForVariable(VariableIdentifier identifier)
+            {
+                int containingSlot = identifier.ContainingSlot;
+                if (containingSlot > 0)
+                {
+                    return GetVariablesForId(DeconstructSlot(containingSlot).Id)!;
+                }
+                return GetVariablesContainingSymbol(identifier.Symbol);
+            }
+
+            private Variables GetVariablesContainingSymbol(Symbol symbol)
+            {
+                switch (symbol)
+                {
+                    case LocalSymbol:
+                    case ParameterSymbol:
+                        if (symbol.ContainingSymbol is MethodSymbol method &&
+                            GetVariablesForMethodScope(method) is { } variables)
+                        {
+                            return variables;
+                        }
+                        break;
+                }
+                // Fallback to the outermost scope for the remaining cases. Those cases include: static fields;
+                // variables declared in field initializers; locals and parameters when the root symbol is null;
+                // and error cases such as an instance field referenced in a static method (no containing slot).
+                return GetRootScope();
+            }
+
+            internal Variables GetRootScope()
+            {
+                var variables = this;
+                while (variables.Container is { } container)
+                {
+                    variables = container;
+                }
+                return variables;
+            }
+
+            private Variables? GetVariablesForId(int id)
+            {
+                var variables = this;
+                do
+                {
+                    if (variables.Id == id)
+                    {
+                        return variables;
+                    }
+                    variables = variables.Container;
+                }
+                while (variables is { });
+                return null;
+            }
+
+            internal Variables? GetVariablesForMethodScope(MethodSymbol method)
+            {
+                method = method.PartialImplementationPart ?? method;
+                var variables = this;
+                while (true)
+                {
+                    if (variables.Symbol is null)
```

Comment:
>if (variables.Symbol is null) [](start = 20, length = 29)

The check feels redundant, it looks like the behavior will be the same without this ```if```.